[Team-19][AOS] [ Linus& 히데 Airbnb 두번째주 금요일 PR ]  인원 선택 페이지 완성,   정보 입력 페이지 구조 수정,   로그인 JWT 받아오기 구현

HyungMinKang opened the pull request:
## 진행사항
- compose 를 활용해 인원 선택 화면의 Header 와 일부 컬럼을 구현
  - GuestSelector와 Divider을 이용해 인원 선택 UI 구현
  - Compose 사용을 위해 라이브러리 추가
  -  Xml과 혼용해서 사용하기 위해 ComposeView를 통해 Compose 구현

- 날짜 입력,  가격범위 입력 페이지  구조 수정
  -  information_input Activity + 공유 뷰모델 형태로 수정
  -  페이지 이동 ,  뒤로가기 시   이전 상태가 유지되도록 수정

- 로그인 JWT 받아오기 
  -  GitHub 로그인후 code를 서버쪽에 넘겨준후 받은 토큰을 저장 

## 결과
![4번째PR (1)](https://user-images.githubusercontent.com/58967292/171786752-5982a3e1-8d0e-405d-bc0f-0e31d4237859.gif)


## 추후 진행예정
- 공유 뷰모델로 구조를 수정하면서  checked버튼과  지우기 넘어가기 버튼이  crash나는 경우가 발생해서 수정할 예정입니다
- 인원 입력 페이지도 공유 뷰모델을 통해 데이터 관리하는 방향으로 수정할 예정입니다 
-  JWT를 정상적으로 받아올 경우  홈화면으로 이동하는 로직을 추가할 예정입니다
-  숙소 검색결과 와 지도에 결과를 표시해주는 화면을 진행할 예정입니다

HyungMinKang committed:
Merge pull request #40 from hwicode/feature/login-view

[Android] [Linus] 로그인 화면 Webview로 구현 완료

HyungMinKang committed:
Merge pull request #43 from hwicode/feature/price-range

[Android] [히데_라이너스] 가격 그래프 화면 구현 완료

kang-proto committed:
Merge branch 'android-develop' of https://github.com/hwicode/airbnb into android

kang-proto committed:
feat:  가격 범위 선택 페이지에서 인원 선택 페이지 이동 구현, ComposeView 사용하기 위해 build.gradle에 라이브러리 추가

- 범위 변경시 지우기 버튼 활성화 -> 지우기 클릭시 차트 범위 초기화
- 넘어가기 클릭시 페이지 이동

cherrylime69 committed:
feat : 인원 선택 화면 일부 compose 로 적용

- compose 를 활용해 인원 선택 화면의 Header 와 일부 컬럼을 구현

kang-proto committed:
feat:  게스트 인원수 페이지 uI 완성,

- GuestSelector와 Divider을 이용해 인원 선택 UI 구현
- To DO: 버튼 클릭 로직 , 클릭에 따른 상단 인원수 변경,  건너뛰기,지우기 로직 적용 필요

cherrylime69 committed:
feat : 인원 선택 화면 compose 로 구현 완료

- rememberSaveable 를 활용하여 버튼을 클릭하면 각 성인, 어린이, 유아 인원 수 카운트하는 기능 구현

cherrylime69 committed:
Merge pull request #57 from hwicode/android-develop

[Android] 브랜치 관리를 위한 PR

kang-proto committed:
Merge branch 'android' of https://github.com/hwicode/airbnb into android

HyungMinKang committed:
Merge pull request #60 from hwicode/feature/guest-range

[Android] [히데] 2주차 첫번째 PR 반영

kang-proto committed:
Merge branch 'android' of https://github.com/hwicode/airbnb into feature/login_oauth

kang-proto committed:
feat: 로그인 jwt 토근을 테스트하기 위한 환경 설정

- DI- Koin 적용
- Retrofit + Gson 직렬화 라이브러리
- API, DataSource 파일 정의

HyungMinKang committed:
Merge pull request #62 from hwicode/feature/login_oauth

[Android]  로그인 OAuth 테스트를 위한  환경 설정

cherrylime69 committed:
feat : 가격 범위 입력 화면 ViewModel 설계로 변경 중

- 체크 버튼, 건너뛰기, 지우기 버튼 로직을 InformationViewModel 을 통해 구현 중

cherrylime69 committed:
feat : 가격 범위 입력 화면 ViewModel 설계로 변경 완료

- index 정보를 통해 다음 화면으로부터 뒤로 가기를 눌러도 저장한 데이터를 화면에 표시하도록 구현 완료

devjun10 committed:
Create ci-dev-be.yml

kang-proto committed:
feat: 로그인 jwt 토근 받아오기

kang-proto committed:
fix: webViewFragment 오타 수정

HyungMinKang committed:
Merge pull request #64 from hwicode/feature/access_token

[Android]  JWT 토큰을 받아오는 환경설정 수정

cherrylime69 committed:
Merge pull request #66 from hwicode/feature/price-guest-viewmodel

[Android][라이너스] 가격 범위 입력 화면 ViewModel 설계로 변경 완료

cherrylime69 committed:
Merge pull request #67 from hwicode/android

[Android] PR 보낼 브랜치 merge

HyungMinKang committed:
docs: make .github dir

HyungMinKang committed:
docs: Upload github template

HyungMinKang committed:
Merge branch 'team-19' into android-develop

kang-proto committed:
Merge branch 'android-develop' of https://github.com/hwicode/airbnb into android-develop

kang-proto committed:
feat: 인원 선택페이지 뷰모델 적용 , 증가,감소 로직 변경

- 성인, 아린이 ,유아 인원수 정보를 뷰모델에서 관리
- 성인이 1명이상일때만 chekc버튼 활성화
- 성인이 0명인 상태에서 유아,어리인이 증가버튼 클릭시 성인1로 변경
- 증가 maxCount 8로 설정

sabgilhun commented on android/app/src/main/java/com/example/airbnb/AirbnbApplication.kt:
```diff
@@ -0,0 +1,19 @@
+
+    override fun onCreate() {
+        super.onCreate()
+        startKoin {
+            androidContext(this@AirbnbApplication)
+            modules(appModule, NetWorkModule)
+        }
```
`Koin`을 사용하셨네요.

`Koin`을 DI가 아닌 Service Locator 패턴이라고 보는 시각들이 있습니다. Service Locator는 어떤 단점이 있는지 아래 글을 봐보시면 좋아요.

https://developer.android.com/training/dependency-injection#di-alternatives

HyungMinKang replied:
Koin의 단점들은 인식하고 있었으나  Service Locator 패턴이라 보는 시각은 처음 들어보았습니다.  좋은 학습거리를 주셔서 감사드립니다. 

Koin을 적용한 이유는 백엔드와 협력과정에서 급하게 DI를 적용할 일이 있어 빠르게 적용할 수 있는 Koin을 사용했습니다.

sabgilhun commented on android/app/src/main/java/com/example/airbnb/common/Constants.kt:
```diff
@@ -2,10 +2,11 @@ package com.example.airbnb.common
 
-    var JSESSIONID = ""
-    const val WEBVIEW_LOGIN_URL = "http://3.34.207.233:8080/login"
+    var CODE = ""
+    var JWT = ""
```
`CODE`, `JWT` 는 상수가 아닌것 같네요.

다른 class를 만들어서 다뤄보는게 좋아보여요.

앱 전역에서 App LIfeCycle과 동일하게 해당 값들이 유지되었으면 한다면 싱글턴 객체 하나 만들어서 해당 값들을 관리하도록 해보는게 어떤가요?

HyungMinKang replied:
별도의  파일로 분리해 관리하는 방향으로 수정했습니다

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/guestRange/GuestRangeFragment.kt:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        skipBtn = view.rootView.findViewById(R.id.btn_information_skip)
+        nextBtn = view.rootView.findViewById(R.id.iBtn_information_next)
+        navigator= Navigation.findNavController(view)
+
+        nextBtn.setOnClickListener {
+            navigator.navigate(R.id.action_guestRangeFragment_to_calendarFragment)
+        }
+        skipBtn.setOnClickListener {
+            navigator.navigate(R.id.action_guestRangeFragment_to_calendarFragment)
+        }
+    }
```
엇 여기 `InformationActivity` SharedViewModel 방식으로 변경되어야 할것 같네요.

HyungMinKang replied:
GuestRange에도  SharedViewModel 적용하도록 수정했습니다.

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/guestRange/GuestRangeFragment.kt:
```diff
@@ -0,0 +1,233 @@
+    }
+}
+
+@Composable
+fun GuessAdultSelectorLayout(countNumber: Int, count: (countNumber: Int) -> Unit) {
+    GuestSelector(type = "성인", ageRange = "만13세이상", countNumber, count)
+}
+
+@Composable
+fun GuessChildSelectorLayout(countNumber: Int, count: (countNumber: Int) -> Unit) {
+    GuestSelector(type = "성인", ageRange = "만13세이상", countNumber, count)
+}
+
+@Composable
+fun GuessToddlerSelectorLayout(countNumber: Int, count: (countNumber: Int) -> Unit) {
+    GuestSelector(type = "성인", ageRange = "만13세이상", countNumber, count)
+}
```
type, agreRange 값 수정 필요해보입니다~

HyungMinKang replied:
넵 수정했습니다

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/guestRange/GuestRangeFragment.kt:
```diff
@@ -0,0 +1,233 @@
+}
+
+@Composable
+fun GuestSelector(type: String, ageRange: String, countNumber: Int, count: (countNumber: Int) -> Unit) {
```
`count` -> `onCountChanged` 이런 네이밍 어떤가요?

HyungMinKang replied:
넵 수정했습니다

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/guestRange/GuestRangeFragment.kt:
```diff
@@ -0,0 +1,233 @@
+            Column(modifier = Modifier
+                .wrapContentHeight()
+                .padding(start = 16.dp, end = 16.dp)) {
+                GuessAdultSelectorLayout(adultCount) {
+                    adultCount = it
+                }
+                Divider(modifier = Modifier
+                    .fillMaxWidth()
+                    .height(1.dp), color = Color(0xFFE0E0E0 ))
+                GuessChildSelectorLayout(childCount) {
+                    childCount = it
+                }
+                Divider(modifier = Modifier
+                    .fillMaxWidth()
+                    .height(1.dp), color = Color(0xFFE0E0E0 ))
+                GuessToddlerSelectorLayout(toddlerCount) {
+                    toddlerCount = it
+                }
```
`GuestSelector` ui 컴포넌트화 좋네요! 👍

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/information/InformationActivity.kt:
```diff
@@ -61,13 +71,17 @@ class InformationActivity : AppCompatActivity() {
+                             viewModel.initFlag()
+                         }
                     }
+                    else -> {}
```
절대 나올 수 없는 상황이라면 

```
error("Invalid fragment ID")
```

이런식으로 error 내버리는 것도 좋습니다.

HyungMinKang replied:
else 분기문이 없어도 된다고 생각해서 수정했습니다

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/login/LoginWebViewFragment.kt:
```diff
@@ -47,14 +54,16 @@ class LoginWebViewFragment : Fragment() {
             super.onPageFinished(view, url)
             Log.d("AppTest", "onPageFinished/ url : ${url}")
+            if (url?.contains("code") == true) {
+                val includeCodeUrl = url?.split("code")?.get(1)
+                includeCodeUrl?.let {
+                    Constants.CODE= (includeCodeUrl.subSequence(1, includeCodeUrl.length)).toString()
+                }
```
Url 에서 code 쿼리 파라미터 값을 꺼내고 싶다면, 아래와 같이 할 수 있습니다.


```
// www.abc.com?code=1234&extra=xxxx
Constants.CODE = Uri.parse(url).getQueryParameter("code") // code = 1234
```

HyungMinKang replied:
넵 수정했습니다.  이전에도 비슷한 방법을 본 것 같은데 까먹고 있었네요,  알려주셔서 감사합니다.

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/information/InformationViewModel.kt:
```diff
@@ -58,37 +80,127 @@ class InformationViewModel:ViewModel() {
+
+    // 여기서부터 PriceRange 관련 메서드
+
+    fun initChart() {
+        viewModelScope.launch {
+            val priceMap = getPriceTestData()
+            val seekBarEntries = ArrayList<BarEntry>()
+
+            // 선 그래프를 막대 그래프로 변경하는 로직
+            for (i in Constants.PRICE_MIN_VALUE..Constants.PRICE_MAX_VALUE) {
+                priceMap[i]?.let {
+                    seekBarEntries.add(BarEntry(i.toFloat(), it.toFloat()))
+                    seekBarEntries.add(BarEntry(i + Constants.SEEKBAR_VALUE_GAP, it.toFloat()))
+                    seekBarEntries.add(
+                        BarEntry(
+                            i + Constants.SEEKBAR_VALUE_GAP,
+                            Constants.SEEKBAR_VACANT_VALUE
+                        )
+                    )
+                    seekBarEntries.add(
+                        BarEntry(
+                            i + Constants.SEEKBAR_VALUE_GAP + Constants.SEEKBAR_VACANT_GAP,
+                            Constants.SEEKBAR_VACANT_VALUE
+                        )
+                    )
+                }
             }
+            _lowestPriceStatedFlow.emit(lowestPrice)
+            _highestPriceStatedFlow.emit(highestPrice)
+            _chartStatedFlow.emit(seekBarEntries)
         }
     }
```
이런 로직들을 ViewModel로 옮긴 것 좋네요 👍

sabgilhun commented:
라이너스, 히데 두분 모두 수고하셨습니다.

전반적으로 UI에 있던 로직이 ViewModel로 많이 옮겨간 것 같아 좋네요.

cherrylime69 committed:
Merge pull request #74 from hwicode/feature/price-guest-viewmodel

[Android] [히데]  인원수 입력 화면 ViewModel 설계로 변경 완료

kang-proto committed:
fix:  두번째주 금요일 PR 리뷰 반영

- JWT, CODE 상수 파일 ->  AccesssToken  파일로 이동
- GuessRangeFragment에 람다 함수명 변경 (count->onCountChanged)
- LoginWebViewFragment의  code parsing 로직 수정
- Inoformation Activity의  분기문 수정

cherrylime69 committed:
Merge pull request #75 from hwicode/fix/secondWeekSecondFeedBack

[Android] 두번째주 금요일 PR 리뷰 반영
